sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1248 @@
+        elif self.activation == 'relu':
+            # He et al instead maybe?
+            init_bound = np.sqrt(6. / (fan_in + fan_out))
+        else:
+            # this was caught earlier, just to make sure
+            raise ValueError("Unknown activation function %s" %
+                             self.activation)
+
+        coef_init = rng.uniform(-init_bound, init_bound, (fan_in, fan_out))
+        intercept_init = rng.uniform(-init_bound, init_bound, fan_out)
+        return coef_init, intercept_init
+
+    def _fit(self, X, y, incremental=False):
+        # Make sure self.hidden_layer_sizes is a list
+        hidden_layer_sizes = self.hidden_layer_sizes
+        if not hasattr(hidden_layer_sizes, "__iter__"):
+            hidden_layer_sizes = [hidden_layer_sizes]
+        hidden_layer_sizes = list(hidden_layer_sizes)
+
+        # Validate input parameters.
+        self._validate_hyperparameters()
+        if np.any(np.array(hidden_layer_sizes) <= 0):
+            raise ValueError("hidden_layer_sizes must be > 0, got %s." %
+                             hidden_layer_sizes)
+
+        X, y = self._validate_input(X, y, incremental)
+        n_samples, n_features = X.shape
+
+        # Ensure y is 2D
+        if y.ndim == 1:
+            y = y.reshape((-1, 1))
+
+        self.n_outputs_ = y.shape[1]
+
+        layer_units = ([n_features] + hidden_layer_sizes +
+                       [self.n_outputs_])
+
+        if not hasattr(self, 'coefs_') or (not self.warm_start and not
+                                           incremental):
+            # First time training the model
+            self._initialize(y, layer_units)
+
+        # l-bfgs does not support mini-batches
+        if self.algorithm == 'l-bfgs':
+            batch_size = n_samples
+        else:
+            batch_size = np.clip(self.batch_size, 1, n_samples)
+
+        # Initialize lists
+        activations = [X]
+        activations.extend(np.empty((batch_size, n_fan_out))
+                           for n_fan_out in layer_units[1:])
+        deltas = [np.empty_like(a_layer) for a_layer in activations]
+
+        coef_grads = [np.empty((n_fan_in_, n_fan_out_)) for n_fan_in_,
+                      n_fan_out_ in zip(layer_units[:-1],
+                                        layer_units[1:])]
+
+        intercept_grads = [np.empty(n_fan_out_) for n_fan_out_ in
+                           layer_units[1:]]
+
+        # Run the Stochastic optimization algorithm
+        if self.algorithm in _STOCHASTIC_ALGOS:
+            self._fit_stochastic(X, y, activations, deltas, coef_grads,
+                                 intercept_grads, layer_units, incremental)
+
+        # Run the LBFGS algorithm
+        elif self.algorithm == 'l-bfgs':
+            self._fit_lbfgs(X, y, activations, deltas, coef_grads,
+                            intercept_grads, layer_units)
+        return self
+
+    def _validate_hyperparameters(self):
+        if not isinstance(self.shuffle, bool):
```

Comment:
Do we really need to check every single hyperparameter for all optimizers here, or can we delegate the relevant checks to the optimizers that implement the algorithms? It would make this a little less intense, but I have no major preference to change what is here. Just looking for opinions.